### PR TITLE
[Linux] Fix Debian 12 auto installation failure

### PR DESCRIPTION
**Pull request by keirazhang**
Debian 12.0 ISO images doesn't have open-vm-tools builtin. So it needs to install open-vm-tools from online repo. In this fix,
1. Added some debug messages in preseed late_command
2. If the Debian ISO image have open-vm-tools builtin, install open-vm-tools from CDROM; otherwise, install it from online repo.
3. Install testing required packages during OS installation.

Test results
```
  - 'Guest os distribution: Debian 10 i386'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian10Guest'
  - 'Guestinfo guest id: debian10Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 10 (32-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''32'' distroAddlVersion=''10
    (buster)'' distroName=''Debian GNU/Linux'' distroVersion=''10'' familyName=''Linux''
    kernelVersion=''4.19.0-21-686-pae'' prettyName=''Debian GNU/Linux 10 (buster)'''
  - 'Results: Total (30), Passed (23), Skipped (7)'
  - ' * deploy_vm_bios_paravirtual_vmxnet3..............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic.....................................Not Supported'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware......................................Not Applicable'
  - ' * secureboot_enable_disable...............................Not Applicable'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp......................................Not Supported'
  - ' * gosc_cloudinit_staticip..................................Not Supported'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove...................................Not Supported'
  - ' * ovt_verify_uninstall............................................Passed'

  - 'Guest os distribution: Debian 10 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian10_64Guest'
  - 'Guestinfo guest id: debian10_64Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 10 (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroAddlVersion=''10
    (buster)'' distroName=''Debian GNU/Linux'' distroVersion=''10'' familyName=''Linux''
    kernelVersion=''4.19.0-21-amd64'' prettyName=''Debian GNU/Linux 10 (buster)'''
  - 'Results: Total (30), Passed (25), Skipped (5)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic............................................Passed'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware..............................................Passed'
  - ' * secureboot_enable_disable................................Not Supported'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp......................................Not Supported'
  - ' * gosc_cloudinit_staticip..................................Not Supported'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove...................................Not Supported'
  - ' * ovt_verify_uninstall............................................Passed'

  - 'Guest os distribution: Debian 11 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian11_64Guest'
  - 'Guestinfo guest id: debian11_64Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 11 (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroAddlVersion=''11
    (bullseye)'' distroName=''Debian GNU/Linux'' distroVersion=''11'' familyName=''Linux''
    kernelVersion=''5.10.0-22-amd64'' prettyName=''Debian GNU/Linux 11 (bullseye)'''
  - 'Results: Total (30), Passed (27), Skipped (3)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic............................................Passed'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware..............................................Passed'
  - ' * secureboot_enable_disable.......................................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp......................................Not Supported'
  - ' * gosc_cloudinit_staticip..................................Not Supported'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove..........................................Passed'
  - ' * ovt_verify_uninstall............................................Passed'

  - 'Guest os distribution: Debian 11 i386'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian11Guest'
  - 'Guestinfo guest id: debian11Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 11 (32-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''32'' distroAddlVersion=''11
    (bullseye)'' distroName=''Debian GNU/Linux'' distroVersion=''11'' familyName=''Linux''
    kernelVersion=''5.10.0-22-686-pae'' prettyName=''Debian GNU/Linux 11 (bullseye)'''
  - 'Results: Total (30), Passed (23), Skipped (7)'
  - ' * deploy_vm_bios_paravirtual_vmxnet3..............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic.....................................Not Supported'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware......................................Not Applicable'
  - ' * secureboot_enable_disable...............................Not Applicable'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp......................................Not Supported'
  - ' * gosc_cloudinit_staticip..................................Not Supported'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove...................................Not Supported'
  - ' * ovt_verify_uninstall............................................Passed'

  - 'Guest os distribution: Debian 12.0 x86_64'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian12_64Guest'
  - 'Guestinfo guest id: debian12_64Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 12 (64-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''64'' distroAddlVersion=''12
    (bookworm)'' distroName=''Debian GNU/Linux'' distroVersion=''12'' familyName=''Linux''
    kernelVersion=''6.1.0-9-amd64'' prettyName=''Debian GNU/Linux 12 (bookworm)'''
  - 'Results: Total (30), Passed (27), Skipped (1), Failed (2)'
  - ' * deploy_vm_efi_paravirtual_vmxnet3...............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic............................................Passed'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware..............................................Passed'
  - ' * secureboot_enable_disable.......................................Passed'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp.............................................Failed'
  - ' * gosc_cloudinit_staticip.........................................Failed'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove..........................................Passed'
  - ' * ovt_verify_uninstall............................................Passed'

  - 'Guest os distribution: Debian 12.0 i386'
  - 'Hardware version: vmx-20'
  - 'Config guest id: debian12Guest'
  - 'Guestinfo guest id: debian12Guest'
  - 'Guestinfo guest full name: Debian GNU/Linux 12 (32-bit)'
  - 'Guestinfo guest family: linuxGuest'
  - 'Guestinfo detailed data: architecture=''X86'' bitness=''32'' distroAddlVersion=''12
    (bookworm)'' distroName=''Debian GNU/Linux'' distroVersion=''12'' familyName=''Linux''
    kernelVersion=''6.1.0-9-686-pae'' prettyName=''Debian GNU/Linux 12 (bookworm)'''
  - 'Results: Total (30), Passed (23), Skipped (5), Failed (2)'
  - ' * deploy_vm_bios_paravirtual_vmxnet3..............................Passed'
  - ' * check_inbox_driver..............................................Passed'
  - ' * ovt_verify_install..............................................Passed'
  - ' * ovt_verify_status...............................................Passed'
  - ' * vgauth_check_service............................................Passed'
  - ' * check_ip_address................................................Passed'
  - ' * check_os_fullname...............................................Passed'
  - ' * stat_balloon....................................................Passed'
  - ' * stat_hosttime...................................................Passed'
  - ' * device_list.....................................................Passed'
  - ' * check_quiesce_snapshot_custom_script............................Passed'
  - ' * memory_hot_add_basic.....................................Not Supported'
  - ' * cpu_hot_add_basic...............................................Passed'
  - ' * cpu_multicores_per_socket.......................................Passed'
  - ' * check_efi_firmware......................................Not Applicable'
  - ' * secureboot_enable_disable...............................Not Applicable'
  - ' * e1000e_network_device_ops.......................................Passed'
  - ' * vmxnet3_network_device_ops......................................Passed'
  - ' * pvrdma_network_device_ops................................Not Supported'
  - ' * gosc_perl_dhcp..................................................Passed'
  - ' * gosc_perl_staticip..............................................Passed'
  - ' * gosc_cloudinit_dhcp.............................................Failed'
  - ' * gosc_cloudinit_staticip.........................................Failed'
  - ' * paravirtual_vhba_device_ops.....................................Passed'
  - ' * lsilogic_vhba_device_ops........................................Passed'
  - ' * lsilogicsas_vhba_device_ops.....................................Passed'
  - ' * sata_vhba_device_ops............................................Passed'
  - ' * nvme_vhba_device_ops............................................Passed'
  - ' * nvdimm_cold_add_remove...................................Not Supported'
  - ' * ovt_verify_uninstall............................................Passed'
```

### CODE REVIEW

File: linux/deploy_vm/deploy_vm_from_iso.yml [hunk 1]
```diff
@@ -182,25 +200,22 @@
...
+    - name: "Wait for guest full name is collected"
+      include_tasks: ../../common/vm_wait_guest_fullname.yml
       when:
         - unattend_install_conf is defined
-        - unattend_install_conf is match('Photon') or unattend_install_conf is match('Ubuntu')
-    
-    - name: "Get VM's primary network adapter MAC address after poweron VM"
-      include_tasks: ../../common/vm_wait_primary_nic_mac.yml
-      when: vm_primary_nic_mac is undefined or not vm_primary_nic_mac
+        - unattend_install_conf | lower is not match('.*(minimal|server_without_gui).*')
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/deploy_vm/deploy_vm_from_iso.yml:207]: It applies to all Linux except when OS is installed with below unattend_install_conf:
```
CentOS/7/minimal/ks.cfg
RHEL/*/server_without_GUI/ks.cfg
SLE/15/SP3/SLES_Minimal/autoinst.xml
```
By using these unattend_install_conf, the installed guest OS doesn't have open-vm-tools. With the other unattend_install_conf files, the installed guest OS should have open-vm-tools installed, so they are expected to have a guest full name.

File: linux/deploy_vm/deploy_vm_from_iso.yml [hunk 2]
```diff
@@ -182,25 +200,22 @@
...
-    # For VMware Photon OS and Ubuntu
-    - include_tasks: ../../common/vm_wait_guest_fullname.yml
+    - name: "Wait for guest full name is collected"
+      include_tasks: ../../common/vm_wait_guest_fullname.yml
       when:
         - unattend_install_conf is defined
-        - unattend_install_conf is match('Photon') or unattend_install_conf is match('Ubuntu')
-    
-    - name: "Get VM's primary network adapter MAC address after poweron VM"
-      include_tasks: ../../common/vm_wait_primary_nic_mac.yml
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/deploy_vm/deploy_vm_from_iso.yml:192]: It's already waited in `common/vm_get_vm_info.yml` after vm creation.

File: linux/deploy_vm/deploy_vm_from_iso.yml [hunk 3]
```diff
@@ -42,76 +45,87 @@
...
+          when: guest_id is match("debian1\\d+")
+
         - name: "Set default unattend install conf file for FreeBSD"
           ansible.builtin.set_fact:
             unattend_install_conf: "FreeBSD/installerconfig"
           when: "'freebsd' in guest_id"
       when: unattend_install_conf is undefined or not unattend_install_conf
 
-    - ansible.builtin.debug:
+    - name: "warning: unattend_install_conf is undefined"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/deploy_vm/deploy_vm_from_iso.yml]: Updated.

File: linux/deploy_vm/deploy_vm_from_iso.yml [hunk 4]
```diff
@@ -182,25 +200,22 @@
...
-        - unattend_install_conf is match('Photon') or unattend_install_conf is match('Ubuntu')
-    
-    - name: "Get VM's primary network adapter MAC address after poweron VM"
-      include_tasks: ../../common/vm_wait_primary_nic_mac.yml
-      when: vm_primary_nic_mac is undefined or not vm_primary_nic_mac
+        - unattend_install_conf | lower is not match('.*(minimal|server_without_gui).*')
 
-    # Get guest IP
-    - include_tasks: ../../common/update_inventory.yml
+    - name: "Get VM guest IPv4 address and add to in-memory inventory"
```
[the remark this answers is not in the document]
Comment (keirazhang) [reply linux/deploy_vm/deploy_vm_from_iso.yml:209]: We are using it in many tasks files. Why can't they be used together?
```
$ grep -rn 'VM guest' common/
common/vm_get_guest_info.yml:42:      - "VM guest info VMware tools info: {{ guestinfo_vmtools_info }}"
common/vm_get_guest_info.yml:43:      - "VM guest info guest id: {{ guestinfo_guest_id }}"
common/vm_get_guest_info.yml:44:      - "VM guest info guest full name: {{ guestinfo_guest_full_name }}"
common/vm_get_guest_info.yml:45:      - "VM guest info guest family: {{ guestinfo_guest_family }}"
common/vm_get_guest_info.yml:46:      - "VM guest info guest detailed data: {{ guestinfo_detailed_data }}"
common/vm_wait_guest_ip.yml:4:# Wait VM guest IP address in guest info is the one specified by parameter 'wait_ipv4'.
common/vm_wait_guest_ip.yml:67:      - "Waiting for VM guest IPv4 address succeeded."
common/vm_get_ip.yml:17:- name: "Initialize VM guest IP address variable"
common/vm_get_id.yml:17:- name: Display the VM guest facts
common/vm_get_guest_facts.yml:15:- name: Display the VM guest facts
common/vm_guest_file_operation.yml:25:    - name: "Create directory {{ dir_path }} in VM guest"
common/vm_guest_file_operation.yml:57:    - name: "Delete directory {{ dir_path }} in VM guest"
common/vm_guest_file_operation.yml:91:    - name: "Fetch file {{ src_path }} from VM guest"
common/vm_guest_file_operation.yml:115:    - name: "Copy file {{ src_path }} to VM guest"
common/vm_get_ip_from_vmtools.yml:17:- name: "Initialize VM guest IP/IP list variables"
common/vm_get_ip_from_vmtools.yml:32:- name: "Get all IP address in VM guest info"
common/vm_get_ip_from_vmtools.yml:51:- name: "Set fact of VM guest IPv4 address with primary network adapter's IPv4 address"
common/vm_set_guest_id.yml:4:# Set VM guest ID
common/vm_set_guest_id.yml:5:- name: "Set VM guest ID to '{{ guest_id }}'"
common/vm_set_guest_id.yml:16:- name: Display changing VM guest ID result
common/vm_wait_guest_hostname.yml:61:- name: "Print VM guest OS hostname"
common/vm_wait_guest_hostname.yml:63:    msg: "Get VM guest OS hostname: {{ vm_guest_facts.instance.guest.hostName }}"
common/README.md:100:* vm_get_vm_info.yml: Get VM datastore path, VM files path, VM guest ID, hardware version info
common/vm_wait_guest_fullname.yml:43:- name: "Print VM guest info"
common/vm_get_vm_info.yml:37:      - "VM guest ID: {{ vm_guest_id }}"
common/vm_get_ip_esxcli.yml:10:#   vm_guest_ip: VM guest IPv4 address
common/vm_get_ip_esxcli.yml:51:- name: "Set fact of VM guest IP with primary IP for VM '{{ vm_name }}'"
```

**Review by 123lzxm — APPROVED** [review]
LGTM, thanks.

**Pull request merged**